Fix bug in IPv4 detection regex

FreedomBen opened the pull request:
The previous regex did not identify IP addresses with a 0 segment as
valid IPv4.  For example, `Iptools.is_ipv4?("127.0.0.1")` would return
`false`.

After fixing the regex, addresses with leading zeroes were now allowed,
for example "08.08.08.08".  To handle this corner case, we add checking
that will catch this and return false as we did before.

Also adds more test cases to really hone behavior.

Fixes #13

NOTE: #10 and possibly #12 should probably be merged prior to this to 
avoid merge conflicts

FreedomBen commented:
If helpful, here are some of the test cases I used for the regex:  https://rubular.com/r/b5FtF8QpqxOQ9M

FreedomBen commented:
I needed this immediately for my app so I've merged all the PRs into a branch and published https://hexdocs.pm/iptools_fb/0.0.7

If anyone wants to test the changes in their app before merging you can use that package:

```elixir
defp deps do
  [
    # ...
    # {:iptools, "~> 0.0.4"},
    {:iptools_fb, "~> 0.0.7"},
  ]
end
```

blackfist commented:
@FreedomBen couple questions before I merge this one - do you think we should also have a `to_list` function which just calls `to_int_list`? I'm worried that someone might be using `to_list` and then an update would remove that function. 

Also, is the typespec for `to_str_list` correct? Should it be `@spec to_str_list(String.t()) :: [integer()]`

blackfist commented:
Similarly the typespec for `to_int_list` seems off. Perhaps it should be `@spec to_int_list(String.t()) :: [integer()]`?

FreedomBen commented:
@blackfist yeah `to_list` is a good idea!  I didn't think about that but it's definitely possible.  And yes it looks like I reversed the typespecs for `to_int_list` and `to_str_list` :laughing:  Will fix

FreedomBen commented:
ok @blackfist I fixed the merge conflicts, the typespecs, and added back in `to_int/1`.  I put a deprecation tag on it but I'm happy to remove it.  Seems like a pretty harmless function to keep around anyway

blackfist commented:
Thanks @FreedomBen - I agree about putting the deprecated tag on `to_list`. We should encourage people to use the most specific function. I've merged the PR and I'll close out the open issue. I'll also publish a version update to hex
